contrib/babelfishpg_tsql/src/collation.c

```
@@ -1639,3 +1666,275 @@ pltsql_replace_non_determinstic(text *src_text, text *from_text, text *to_text,
+							NextChar(p, plen);
+						}
+						
+						src_text = cstring_to_text_with_len(p_start, p_start_len - plen);
+						substr_text = cstring_to_text_with_len(t, pg_mblen(t));
+						
+						if (pltsql_strpos_non_determinstic(src_text, substr_text, cid, &matched_idx) && matched_idx != 0)
+						{
+							find_match = true;
+						}
+						pfree(src_text);
+						pfree(substr_text);
+					}
+				}
+
+				if (close_bracket && (find_match ^ reverse_mode)) /* found a match */
+					NextChar(t, tlen);
+				else
+				{
+					match_failed = true;
+					break;
+				}
+			}
+			else
+			{
+				char *p_start = p;
+				int  len = plen, matched_len = 0;
+
+				while (plen > 0 && *p != '[' && *p != '_' && *p != '%')
+				{
+					NextChar(p, plen);
+				}
+				if (icu_find_matched_length(t, strlen(t), p_start, len-plen, cid, &matched_len, is_cs_ai))
+				{
+					while (matched_len--)
+						NextChar(t, tlen);
+				}
+				else
+				{
+					match_failed = true;
+					break;
+				}
+			}
+		}
+
+		if (tlen > 0 && !match_failed)
+		{
+			while (tlen > 0 && *t == ' ')
+				NextChar(t, tlen);
+			if (tlen <= 0 && plen <=0)
+				return itr;
+		}
+
+		/*
+		* End of text, but perhaps not of pattern.  Match if the remaining
+		* pattern can match a zero-length string, ie, it's zero or more %'s.
+		*/
+		while (plen > 0 && *p == '%')
+			NextChar(p, plen);
+		if (plen <= 0 && !match_failed)
+			return itr;
+
+		if (start_offset)
+			input_str += pg_mblen(input_str);
+		else
+			break;
+	}
+
+	return 0;
+}
+
+PG_FUNCTION_INFO_V1(patindex_ai_collations);
+Datum
+patindex_ai_collations(PG_FUNCTION_ARGS)
+{
+	pg_locale_t  mylocale = 0;
+	char         *pattern = text_to_cstring(PG_GETARG_TEXT_PP(0));
+	char         *input_str = text_to_cstring(PG_GETARG_TEXT_PP(1));
+	Oid          cid = PG_GET_COLLATION();
+	int 		 result = 0;
+	bool		 is_CS_AI = false;
+	coll_info_t  coll_info_of_inputcollid = tsql_lookup_collation_table_internal(cid);
+
+	if (OidIsValid(coll_info_of_inputcollid.oid) &&
```

We should throw an error here if `coll_info_of_inputcollid.oid == InvalidOid`